django/db/models/expressions.py

```diff
@@ -843,13 +843,19 @@ def as_sql(self, compiler, connection, template=None, case_joiner=None, **extra_
         template_params.update(extra_context)
         case_parts = []
         sql_params = []
+        from django.db.models.sql.datastructures import EmptyResultSet
```

Comment:
Yes there is a chance it may be needed by future expressions. The subquery expression that @schinckel is putting together may end up needing it too. Purity is also a reason though. The error isn't really sql specific so it shouldn't be pushed down into the sql implementation - much nicer to have it somewhere in the frontend. History tells me it used to live in `django.db.models.query` until it was moved into `django.db.models.sql.query` during the queryset refactor - but I think that was just because it was used only within the sql module at that point.